lib/pbench/test/unit/server/test_requests.py

```diff
@@ -170,3 +175,204 @@ def test_upload(client, pytestconfig, caplog):
+        assert response.status_code == 400
+        assert (
+            response.json.get("message") == "QueryControllers: Missing request payload"
+        )
+
+    @pytest.mark.parametrize(
+        "keys",
+        (
+            {"user": "x"},
+            {"start": "2020"},
+            {"end": "2020"},
+            {"user": "x", "start": "2020"},
+            {"user": "x", "end": "2020"},
+            {"start": "2020", "end": "2020"},
+        ),
+    )
+    def test_missing_keys(self, client, keys):
+        """
+        test_missing_keys Test behavior when JSON payload does not contain
+        all required keys.
+
+        Note that "user", "prefix", "start", and "end" are all required;
+        however, Pbench will silently ignore any additional keys that are
+        specified.
+       """
+        response = client.post(
+            f"{client.config['REST_URI']}/controllers/list", json=keys
+        )
+        assert response.status_code == 400
+        missing = [k for k in ("user", "start", "end") if k not in keys]
+        assert (
+            response.json.get("message")
+            == f"QueryControllers: Missing request data: {','.join(missing)}"
+        )
+
+    def test_bad_dates(self, client):
+        """
+        test_bad_dates Test behavior when a bad date string is given
+        """
+        response = client.post(
+            f"{client.config['REST_URI']}/controllers/list",
+            json={
+                "user": "drb",
+                "prefix": "drb-",
+                "start": "2020-15",
+                "end": "2020-19",
+            },
+        )
+        assert response.status_code == 400
+        assert (
+            response.json.get("message")
+            == "QueryControllers: Invalid start or end time string"
+        )
+
+    def test_query(self, client, query_helper):
+        """
+        test_query Check the construction of Elasticsearch query URI
+        and filtering of the response body.
+        """
+        json = {
+            "user": "drb",
+            "start": "2020-08",
+            "end": "2020-10",
+        }
+        response_payload = {
+            "took": 1,
+            "timed_out": False,
+            "_shards": {"total": 1, "successful": 1, "skipped": 0, "failed": 0},
+            "hits": {
+                "total": {"value": 2, "relation": "eq"},
+                "max_score": None,
+                "hits": [],
+            },
+            "aggregations": {
+                "controllers": {
+                    "doc_count_error_upper_bound": 0,
+                    "sum_other_doc_count": 0,
+                    "buckets": [
+                        {
+                            "key": "unittest-controller1",
+                            "doc_count": 2,
+                            "runs": {
+                                "value": 1.59847315581e12,
+                                "value_as_string": "2020-08-26T20:19:15.810Z",
```

Comment:
why is `value_as_string` a datetime vale?

Reply:
Because it's the highest value of the selected aggregate  `"aggs": {"runs": {"max": {"field": "run.start"}}}` in a set of controllers found by the search query. Elasticsearch represents the result in both binary and string form, and the existing Dashboard model picks those out so I needed to put them in my sample response.